python/ray/autoscaler/aws/cloudwatch/cloudwatch_helper.py

```diff
@@ -0,0 +1,257 @@
+        # wait for all EC2 instance checks to complete
+        try:
+            waiter = self.ec2_client.get_waiter("instance_status_ok")
+            waiter.wait(InstanceIds=self.node_ids)
+        except botocore.exceptions.WaiterError as e:
+            logger.error(
+                "Failed while waiting for EC2 instance checks to complete: {}".
+                format(e.message))
+            raise e
+
+        # install the cloudwatch agent on each cluster node
+        parameters_cwa_install = {
+            "action": ["Install"],
+            "name": ["AmazonCloudWatchAgent"],
+            "version": ["latest"]
+        }
+        response = self._send_command_to_all_nodes(
+            "AWS-ConfigureAWSPackage",
+            parameters_cwa_install,
+        )
+        command_id = response["Command"]["CommandId"]
+        try:
+            self._ssm_command_waiter(command_id)
+            logger.info(
+                "SSM successfully installed cloudwatch agent on {} nodes".
+                format(len(self.node_ids)))
+        except botocore.exceptions.WaiterError as e:
+            logger.error(
+                "{}: Failed while waiting for SSM command to complete "
+                "on all cluster nodes".format(e))
+            raise e
+
+        # upload cloudwatch agent config to the SSM parameter store
+        cwa_config = self._load_config_file("agent")
+        self._replace_all_config_variables(
+            cwa_config,
+            str(self.node_ids),
+            self.cluster_name,
+        )
+        self.ssm_client.put_parameter(
+            Name=CWA_CONFIG_SSM_PARAM_NAME,
+            Type="String",
+            Value=json.dumps(cwa_config),
+            Overwrite=True,
+        )
+
+        # satisfy collectd preconditions before installing cloudwatch agent
+        parameters_run_shell = {
+            "commands": [
+                "mkdir -p /usr/share/collectd/",
+                "touch /usr/share/collectd/types.db"
+            ],
+        }
+        self._send_command_to_all_nodes(
+            "AWS-RunShellScript",
+            parameters_run_shell,
+        )
+
+        # start cloudwatch agent
+        parameters_start_cwa = {
+            "action": ["configure"],
+            "mode": ["ec2"],
+            "optionalConfigurationSource": ["ssm"],
+            "optionalConfigurationLocation": [CWA_CONFIG_SSM_PARAM_NAME],
+            "optionalRestart": ["yes"],
+        }
+        self._send_command_to_all_nodes(
+            "AmazonCloudWatch-ManageAgent",
+            parameters_start_cwa,
+        )
+
+    def put_cloudwatch_dashboard(self):
+        """put dashboard to cloudwatch console"""
+
+        cloudwatch_config = self.provider_config["cloudwatch"]
+        dashboard_config = cloudwatch_config.get("dashboard", {})
+        dashboard_name = dashboard_config.get("name", self.cluster_name)
+        data = self._load_config_file("dashboard")
+        widgets = []
+        for node_id in self.node_ids:
```

Comment:
Putting a dashboard with the same name for each instance ID doesn't seem desirable. Currently there is only one dashboard that can be referenced from config, so that implies one CloudWatch dashboard per Ray cluster. 

Having a single dashboard for the entire cluster also precludes referencing instance IDs within the dashboard config, since calls to put_dashboard doesn't take the symmetric difference of widgets between two calls, but always replaces the dashboards widgets with those specified in the latest call.

I see a few ways to resolve this issue:
1) Only support one dashboard for each cluster initially, and remove {instance_id} variable support from it.
2) Support a set of dashboard configs for the cluster, allowing both the dashboard name and config to reference per-EC2-instance variables.

I think it's probably ok to start with support for (1) in this PR, and then follow-up with an appropriate design for (2) in a subsequent PR.